cookbooks: create an AI cookbook for RAG (#44)
redpanda-data/rp-connect-docs: modules/cookbooks/pages/rag.adoc

> = Retrival Augmented Generation (RAG)

Reviewer: ```suggestion
= Retrieval-Augmented Generation (RAG)
```

// tag::single-source[]

> :description: How to configure Redpanda Connect to create a RAG pipeline using Postgres and PGVector.

Reviewer: ```suggestion
:description: How to configure Redpanda Connect to create a RAG pipeline, using Postgres and PGVector.
```

> This cookbook demonstrates how to create a Retrival Augmented Generation (RAG) pipeline using Postgres and https://github.com/pgvector/pgvector[PGVector]. We'll create two Redpanda Connect pipelines: one for creating embeddings with https://ollama.ai[Ollama^] and another for searching the data using https://github.com/pgvector/pgvector[PGVector]. We will run both pipelines together using xref:guides/streams_mode/about.adoc[streams mode].

Reviewer: ```suggestion
This cookbook shows you how to create a Retrieval-Augmented Generation (RAG) pipeline, using Postgres and https://github.com/pgvector/pgvector[PGVector^]. 

Follow the cookbook to:

* Create two Redpanda Connect data pipelines: one for creating embeddings with https://ollama.ai[Ollama^] and another for searching the data using https://github.com/pgvector/pgvector[PGVector]
* Run the two pipelines in parallel using xref:guides/streams_mode/about.adoc[streams mode].
```

Author: > Are there any prerequisites?

No `http` components, did I accidentally add that in anywhere? The only prerequisite is postgres running somewhere that has PGVector enabled.

== Compute the embeddings

> We'll start out by defining the indexing pipeline that will take textual data from Redpanda, compute embeddings for it, and then write it into a Postgres table with a PGVector index on the embeddings column.

Reviewer: ```suggestion
Start by defining the indexing pipeline, which takes textual data from Redpanda, computes embeddings for it, and then writes it into a Postgres table with a PGVector index on the embeddings column.
```
When you say Redpanda, specifically what do you mean? A Redpanda topic on a self-managed/Cloud cluster? Do we need to spell out what embeddings are here first?

Author: Yes sounds good to me! I used the phrase `vector embeddings`

Reviewer: Can we still include original suggestion above, to change the text from. We'll start out to.... Start by defining?

Author: Yes sorry, didn't mean to revert.

> Articles to be indexed are received over the topic `articles` and look like this:

Reviewer: I'm not sure what you mean when you say `received over`.  I guess I'm not sure what the source of this json is but that maybe my lack of knowledge.

> Articles to be indexed are received over the topic `articles` and look like this:

Reviewer: ```suggestion
Articles for indexing are received over the topic `articles` and look like this:
```

[source,json]
----

> {

Reviewer: Do we need to talk about where this data is created?

Reviewer: I think, that again assumes an experienced user. So if you want to lower the barrier, you could add it in?

Author: Added it in

  "type": "article",
  "article": {
    "id": "123foo",
    "title": "Dogs Stop Barking",
    "content": "The world was shocked this morning to find that all dogs have stopped barking."
  }
}
----

> We'll be able to read from our Redpanda topic using the xref:components:inputs/kafka.adoc[`kafka`] input:

Reviewer: ```suggestion
The indexing pipeline reads from the Redpanda topic using the xref:components:inputs/kafka.adoc[`kafka`] input:
```

[source,yaml]
----
input:
  kafka:
    addresses: [ "${REDPANDA_CLUSTER}" ]
    topics: [ articles ]
    consumer_group: rp_connect_articles_group
    tls:
      enabled: true
    sasl:
      mechanism: SCRAM-SHA-256
      user: "${REDPANDA_USER}"
      password: "${REDPANDA_PASSWORD}"
----

> We'll be using https://ollama.com/library/nomic-embed-text[Nomic Embed] to compute embeddings. Since each request only applies to a single document we will make this scale by making requests in parallel across our document batches.

Reviewer: ```suggestion
You can use https://ollama.com/library/nomic-embed-text[Nomic Embed] to compute embeddings. Since each request only applies to a single document, you can scale this by making requests in parallel across our document batches.
```

In order to send a mapped request and map the response back into the original document we will use the xref:components:processors/branch.adoc[`branch` processor], with a child xref:components:processors/ollama_embeddings.adoc[`ollama_embeddings`] processor.

[source,yaml]
----
pipeline:
  threads: -1
  processors:
    - branch:
        request_map: 'root = "search_document: %s\n%s".format(this.article.title, this.article.content)'
        processors:
          - ollama_embeddings:
              model: nomic-embed-text
        result_map: 'root.article.embeddings = this'
----

> With this pipeline our documents will come out looking something like this:

Reviewer: ```suggestion
With this pipeline, your processed documents should look something like this:
```

----
{
  "type": "article",
  "article": {
    "id": "123foo",
    "title": "Dogs Stop Barking",
    "content": "The world was shocked this morning to find that all dogs have stopped barking.",
    "embeddings": [0.754, 0.19283, 0.231, 0.834], # This vector will actually have 768 dimensions
  }
}
----

Now we can send that to Postgres using the xref:components:outputs/sql_insert.adoc[`sql_insert`] output. We'll also take advantage of the `init_statement` functionality to setup our `pgvector` and our table.

[source,yaml]
----
output:
  sql_insert:
    driver: postgres
    dsn: "${PG_CONNECTION_STRING}"
    init_statement: |
      CREATE EXTENSION IF NOT EXISTS vector;
      CREATE TABLE IF NOT EXISTS searchable_text (
        id varchar(128) PRIMARY KEY,
        title text NOT NULL,
        body text NOT NULL,
        embeddings vector(768) NOT NULL
      );
      CREATE INDEX IF NOT EXISTS text_hnsw_index
        ON searchable_text 
        USING hnsw (embeddings vector_l2_ops);
    table: searchable_text
    columns: ["id", "title", "body", "embeddings"]
    args_mapping: "[this.article.id, this.article.title, this.article.content, this.article.embeddings.vector()]"
----

> We can save this pipeline as `indexing.yaml` and run it with `rpk connect run ./indexing.yaml` to test and make sure our postgres table is being populated with embeddings.

Reviewer: ```suggestion
Save this pipeline as `indexing.yaml` and run it with `rpk connect run ./indexing.yaml` to make sure your PostgreSQL table is populated with embeddings.
```

== Generating search responses

> In order to generate responses to questions over our dataset we're creating embeddings for, we'll use an xref:components:inputs/http_server.adoc[HTTP server input] to recieve the questions. Since we'll be running this pipeline in Streams mode, our HTTP server will xref:guides:streams_mode/about.adoc#http-endpoints[endpoints prefixed with stream identifier], which in our case is the same as the filename: `search`.

Reviewer: ```suggestion
To generate responses to questions with the dataset you're creating embeddings for, use a xref:components:inputs/http_server.adoc[HTTP server input] to receive the questions. Since you're running this pipeline in streams mode, the HTTP server uses xref:guides:streams_mode/about.adoc#http-endpoints[endpoints prefixed with stream identifier], which in this case is the same as the filename: `search`.
```

[source,yaml]
----
input:
  http_server:
    path: /
    allowed_verbs: [ GET ]
    sync_response:
      headers:
        Content-Type: application/json
----

Our user query will come in as the query parameter `q` like so `http://localhost:4195/search?q=question_here`. Since query parameters as exposed as metadata in the `http_server` input, we can reference that in bloblang with `@q`.

[source,yaml]
----
pipeline:
  processors:
    - label: compute_embeddings
      ollama_embeddings:
        model: nomic-embed-text
        text: "search_query: %s".format(@q)
----

> Now our payload is the embeddings vector and we can use Postgres to fetch the top 3 most similar documents to our embeddings vector using the xref:components:processors/sql_raw.adoc[`sql_raw`] processor.

Reviewer: ```suggestion
Now, the payload is the embeddings vector, and you can use PostgreSQL to fetch the top three most similar documents to our embeddings vector with the xref:components:processors/sql_raw.adoc[`sql_raw`] processor.
```

[source,yaml]
----
    - sql_raw:
        driver: "postgres"
        dsn: "${PG_CONNECTION_STRING}"

>         query: SELECT title, body FROM searchable_text ORDER BY embeddings <-> $1 LIMIT 3

Reviewer: The <-> notation is showing as an editable field in the preview. You can switch off this functionality in the snippet by adding this inside the start of the code block. [,yaml, role="no-placeholders"].

        args_mapping: "[ this.vector() ]"
----

With our looked up information, as well as our initial query, we can use the xref:components:processors/ollama_chat.adoc[`ollama_chat` processor] to respond to the user's question as text.

[source,yaml]
----
    - label: generate_response
      ollama_chat:
        model: llama3.1
        prompt: |
          Your task is to respond to user queries using the provided information.

          The user asked: ${! @q }
          Context: ${!this.map_each(row -> "%s\n%s".format(row.title, row.body)).join("\n\n")}
          Response:
----

> Now that we've generated our response, we can send that back as the response to our HTTP server using xref:components:processors/sync_response.adoc[`sync_response`], then we will delete the message so nothing goes to the output using a xref:components:processors/mapping.adoc[bloblang mapping].

Reviewer: ```suggestion
Now that you've generated a response, you can send that back to the to our HTTP server as a response using xref:components:processors/sync_response.adoc[`sync_response`]. Then, delete the message using a xref:components:processors/mapping.adoc[bloblang mapping] so that nothing goes to the output.
```

[source,yaml]
----
    - mapping: 'root.response = content().string()'
    - sync_response: {}
    - mapping: 'root = deleted()'
----

> Now that that both pipelines are ready, we can run them both using streams mode: `rpk connect streams indexing.yaml search.yaml`.

Reviewer: ```suggestion
Both pipelines are ready. Try running both of them using streams mode: `rpk connect streams indexing.yaml search.yaml`.
```

> Once some documents have been indexed, we can query our system via: `curl -G 'localhost:4195/search' --data-urlencode 'q=what is happening to the dogs?' | jq`

Reviewer: ```suggestion
Once some documents have been indexed, you can query the system using: `curl -G 'localhost:4195/search' --data-urlencode 'q=what is happening to the dogs?' | jq`
```

> The output will look something like:

Reviewer: ```suggestion
The output should look something like:
```

Follow the cookbook to:

* Create two Redpanda Connect data pipelines: one for creating embeddings with https://ollama.ai[Ollama^] and another for searching the data, using https://github.com/pgvector/pgvector[PGVector^].
* Run the two pipelines in parallel using xref:guides/streams_mode/about.adoc[streams mode].

== Compute the embeddings

We'll start out by defining the indexing pipeline that will take textual data from Redpanda, compute embeddings for it, and then write it into a Postgres table with a PGVector index on the embeddings column.

Articles for indexing are received over the topic `articles` and look like this:

[source,json]
----
{
  "type": "article",
  "article": {
    "id": "123foo",
    "title": "Dogs Stop Barking",
    "content": "The world was shocked this morning to find that all dogs have stopped barking."
  }
}
----

Your indexing pipeline can read from the Redpanda topic, using the xref:components:inputs/kafka.adoc[`kafka`] input:

[source,yaml]
----
input:
  kafka:
    addresses: [ "${REDPANDA_CLUSTER}" ]
    topics: [ articles ]
    consumer_group: rp_connect_articles_group
    tls:
      enabled: true
    sasl:
      mechanism: SCRAM-SHA-256

>       user: "${REDPANDA_USER}"

Author: Good point. For cloud should I just add TODO like we do in the other cookbooks?

>       user: "${REDPANDA_USER}"

Author: I did that, PTAL.

= Retrieval-Augmented Generation (RAG)
// tag::single-source[]

> :description: How to configure Redpanda Connect to create a RAG pipeline, using PostgreSQL and PGVector.

Reviewer 2: Let's not single-source the description. We may want to customize it for cloud one day.

Author: Sorry what is the suggestion? Put the single-source tag below the description or `ifdef` the description? Sorry I'm not very familar with how all these steps get rendered.

Reviewer: To put the // tag::single-source[] below the description.

Author: Done

ifndef::env-cloud[]
This cookbook shows you how to create a Retrieval-Augmented Generation (RAG) pipeline, using PostgreSQL and https://github.com/pgvector/pgvector[PGVector^]. 

Follow the cookbook to:

* Create two Redpanda Connect data pipelines: one for creating vector embeddings with https://ollama.ai[Ollama^] and another for searching the data, using https://github.com/pgvector/pgvector[PGVector^].

> * Run the two pipelines in parallel using xref:guides/streams_mode/about.adoc[streams mode].

Reviewer: ```suggestion
* Run the two pipelines in parallel using xref:guides:streams_mode/about.adoc[streams mode].
```
This is a broken link at the moment.

Author: Ah thanks! Fixed

endif::[]

== Compute the embeddings

ifndef::env-cloud[]
Start by defining the indexing pipeline, which takes textual data from Redpanda, computes vector embeddings for it, and then write it into a PostgreSQL table with a PGVector index on the embeddings column.
endif::[]

[source,bash]
----
rpk topic create articles
echo '{
  "type": "article",
  "article": {
    "id": "123foo",
    "title": "Dogs Stop Barking",
    "content": "The world was shocked this morning to find that all dogs have stopped barking."
  }
}' | rpk topic produce acticles -f '%v'
----

Your indexing pipeline can read from the Redpanda topic, using the xref:components:inputs/kafka.adoc[`kafka`] input:

[source,yaml]
----
input:
  kafka:
ifndef::env-cloud[]
    addresses: [ "${REDPANDA_CLUSTER}" ]
endif::[]
ifdef::env-cloud[]
    addresses: [ "TODO" ]
endif::[]
    topics: [ articles ]
    consumer_group: rp_connect_articles_group
    tls:
      enabled: true
    sasl:
      mechanism: SCRAM-SHA-256
ifndef::env-cloud[]
      user: "${REDPANDA_USER}"
      password: "${REDPANDA_PASSWORD}"
endif::[]
ifdef::env-cloud[]
      user: "TODO"
      password: "TODO"
endif::[]
----

> Use https://ollama.com/library/nomic-embed-text[Nomic Embed] to compute embeddings. Since each request only applies to a single document, you can scale this by making requests in parallel across document batches.

Reviewer 2: ```suggestion
Use https://ollama.com/library/nomic-embed-text[Nomic Embed^] to compute embeddings. Since each request only applies to a single document, you can scale this by making requests in parallel across document batches.
```

[source,yaml]
----
input:
  http_server:
    path: /
    allowed_verbs: [ GET ]
    sync_response:
      headers:
        Content-Type: application/json
----

The user query uses the query parameter `q`: `http://localhost:4195/search?q=question_here`. Since query parameters are exposed as metadata in the `http_server` input, you can reference that in bloblang with `@q`.

[source,yaml]
----
pipeline:
  processors:
    - label: compute_embeddings
      ollama_embeddings:
        model: nomic-embed-text
        text: "search_query: %s".format(@q)
----

The payload is the embeddings vector. To fetch the top three most similar documents to your embeddings vector, you can write a PostgreSQL query in the xref:components:processors/sql_raw.adoc[`sql_raw`] processor.

[source,yaml,role="no-placeholders"]
----
    - sql_raw:
        driver: "postgres"
        dsn: "${PG_CONNECTION_STRING}"
        query: SELECT title, body FROM searchable_text ORDER BY embeddings <-> $1 LIMIT 3
        args_mapping: "[ this.vector() ]"
----

> With our looked up information, as well as our initial query, we can use the xref:components:processors/ollama_chat.adoc[`ollama_chat` processor] to respond to the user's question as text.

Reviewer: ```suggestion
With your looked up information, as well as your initial query, you can use the xref:components:processors/ollama_chat.adoc[`ollama_chat` processor] to respond to the user's question as text.
```

[source,yaml]
----
    - label: generate_response
      ollama_chat:
        model: llama3.1
        prompt: |
          Your task is to respond to user queries using the provided information.

          The user asked: ${! @q }
          Context: ${!this.map_each(row -> "%s\n%s".format(row.title, row.body)).join("\n\n")}
          Response:
----

Now that you've generated a response, you can send that back to the HTTP server as a response using xref:components:processors/sync_response.adoc[`sync_response`]. Then, delete the message using a xref:components:processors/mapping.adoc[bloblang mapping] so that nothing goes to the output.

[source,yaml]
----
    - mapping: 'root.response = content().string()'
    - sync_response: {}
    - mapping: 'root = deleted()'
----

Both pipelines are ready. Try running both of them using streams mode: `rpk connect streams indexing.yaml search.yaml`.

> Once some documents have been indexed, you can query the system using: `curl -G 'localhost:4195/search' --data-urlencode 'q=what is happening to the dogs?' | jq`

Reviewer 2: ```suggestion
When some documents have been indexed, you can query the system using:

[,bash]
----
curl -G 'localhost:4195/search' --data-urlencode 'q=what is happening to the dogs?' | jq
----
```

The output should look something like:

[source,json]
----
{

>   "response": "Everyone in the world woke up today shocked as their beloved pooches were slient - unable to bark."

Reviewer 2: ```suggestion
  "response": "Everyone in the world woke up today shocked as their beloved pooches were silent - unable to bark."
```
